packages/libs/pactjs-generator/src/contract/parsing/pactParser.ts

```diff
@@ -0,0 +1,410 @@
+      : [];
+
+  return [extModules, pointer];
+}
+
+interface IModuleLoader {
+  getModule(moduleFullName: string): Promise<IModuleWithPointer | undefined>;
+  parserFile(content: string): void;
+  getStorage(): Map<string, IModuleWithPointer>;
+}
+
+const moduleLoader = (
+  fetchModule: (moduleFullName: string) => Promise<string>,
+): IModuleLoader => {
+  const storage = new Map<string, IModuleWithPointer>();
+
+  const parseModule = (content: string, namespace?: string): void => {
+    const [modules, pointer] = fileParser(content, namespace);
+    modules.forEach((mod) => {
+      storage.set(getModuleFullName(mod), { ...mod, pointer });
+    });
+  };
+
+  const cache = new Map<string, Promise<string>>();
+  const fetchContract = (name: string): Promise<string> => {
+    if (cache.has(name)) return cache.get(name)!;
+    const pr = fetchModule(name);
+    cache.set(name, pr);
+    return pr;
+  };
+
+  return {
+    async getModule(moduleFullName: string) {
+      const slogs = moduleFullName.split('.');
+      const namespace = slogs.length === 2 ? slogs[0] : '';
+
+      const mod = storage.get(moduleFullName);
+      if (mod !== undefined) return mod;
+
+      const content = await fetchContract(moduleFullName);
+      if (content) {
+        parseModule(content, namespace);
+      }
+
+      return storage.get(moduleFullName)!;
+    },
+    parserFile(content: string) {
+      parseModule(content);
+    },
+    getStorage() {
+      return storage;
+    },
+  };
+};
+
+interface IModuleWithPointer extends IModule {
+  pointer?: IPointer;
+}
+
+// parse a module and its dependencies
+async function loadModuleDependencies(
+  fullName: string,
+  getModule: (name: string) => Promise<IModuleWithPointer | undefined>,
+): Promise<Array<IModuleWithPointer>> {
+  const module = await getModule(fullName);
+  // skip this module if its not available
+  if (!module) return [];
+
+  const interfaceOrModule = [
+    ...(module.usedModules || []),
+    ...(module.usedInterface || []),
+  ];
+
+  const mods = await Promise.all(
+    interfaceOrModule.map(async (usedModule) =>
+      loadModuleDependencies(getModuleFullName(usedModule), getModule),
+    ),
+  );
+  const dependencies = mods.flat();
+
+  return [module, ...dependencies].filter(isNotDuplicatedModule);
+}
+
+// check function body again for checking function calls, internal and external
+const getFunctionCalls = (
+  bodyPointer: number | undefined,
+  pointer: IPointer,
+  fnList: string[],
+  availableExternals: Array<{
+    namespace?: string;
+    module: string;
+    func: string;
+  }>,
+): {
+  internal?: string[];
+  external?: Array<{ namespace?: string; module: string; func: string }>;
+} => {
+  if (bodyPointer !== undefined) {
+    pointer.reset(bodyPointer);
+    const blockPointer = getBlockPointer(pointer);
+    const data = functionCalls(fnList, availableExternals)(blockPointer);
+    if (data !== FAILED) {
+      return unwrapData(data);
+    }
+  }
+
+  return {};
```

Comment:
Can you invert the if statement so it'll be an early return?

```suggestion
  if (bodyPointer === undefined) {
    return {};
  }
  
  pointer.reset(bodyPointer);
  const blockPointer = getBlockPointer(pointer);
  const data = functionCalls(fnList, availableExternals)(blockPointer);
  if (data !== FAILED) {
    return unwrapData(data);
  }
```